### PR TITLE
GOVSI-683: Implement updated T&Cs state

### PR DESCRIPTION
**Pull request by chrisclayson**
## What?

- Add a fluent state machine builder API
- Add a conditional path that checks whether or not the user has accepted the latest T&Cs, if not, changes to the appropriate state
- Add missing StateMachine unit tests.

## Why?

We have several conditional paths in our user journey, this PR allows this functionality in the state machine and adds one conditional state to demonstrate how this achieved.

### CODE REVIEW

File: shared/src/main/java/uk/gov/di/authentication/shared/state/StateMachine.java [hunk 1]
```diff
@@ -74,148 +82,144 @@ public T transition(T from, A action) {
...
+                            .collect(Collectors.toList()));
+            return stateMachineBuilder;
+        }
+
+        protected StateRuleBuilder(Builder<T, A, C> stateMachineBuilder, T state) {
+            this.stateMachineBuilder = stateMachineBuilder;
+            this.state = state;
+        }
+
+        public Builder<T, A, C> deadEnd() {
```
Comment (sanisc) [inline shared/src/main/java/uk/gov/di/authentication/shared/state/StateMachine.java:217]: could this be changed to final?

Comment (chrisclayson) [reply shared/src/main/java/uk/gov/di/authentication/shared/state/StateMachine.java]: `final` is a reserved word, so had to use `finalState()`

File: shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java [hunk 2]
```diff
@@ -5,6 +5,15 @@
 
 public class ConfigurationService {
 
+    private static ConfigurationService configurationService;
+
+    public static ConfigurationService getInstance() {
```
Comment (JHjava) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:10]: Is this used anywhere?

Comment (chrisclayson) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:10]: Good spot ... not used any more, changed my mind on how to inject the config service.

Comment (chrisclayson) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:10]: Yes ... in `StateMachine.java`, line 106 ... it's a large diff, so may not show by default.

Comment (JHjava) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:10]: Ah yes that will be it

**Review by JHjava — APPROVED** [review]

**Pull request merged**